---
title: "HiSeq Report"
author: "Ming Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
    keep_md: false
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  input_dir: ""
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(knitr))

knitr::opts_chunk$set(fig.width  = 12, 
                      fig.height = 8, 
                      fig.path   = "Figures/",
                      eval       = TRUE,
                      echo       = FALSE,
                      cache      = FALSE,
                      prompt     = FALSE,
                      tidy       = FALSE,
                      comment    = NA,
                      message    = FALSE,
                      warning    = FALSE,
                      eval       = T,
                      rownames.print = FALSE)
```

```{r}
library(hiseqr)
library(ggplot2)
library(patchwork)
library(fishualize)
library(dplyr)
library(glue)
library(ggcor)
library(GenomicRanges)
library(patchwork)
```

```{r config, eval = T}
input_dir  <- params$input_dir
input_dir  <- normalizePath(input_dir)
hiseq_type <- get_hiseq_type(input_dir)
if(! is_hiseq_single_dir(input_dir)) {
  stop(paste0("Not a RNAseq single directory, _r1: ", input_dir))
}
```

```{r}
pd       <- hiseqr::read_hiseq(input_dir)
organism <- pd$args$genome
genome   <- pd$args$genome
gtf      <- pd$args$gtf
ctl_name <- pd$args$wildtype_name
exp_name <- pd$args$mutant_name
project_name <- pd$args$project_name

# smp merge list
rep_list <- list_rnaseq_single_dirs(input_dir)

# list of files
df_trim_stat  <- get_rnaseq_trim_stat(input_dir)
df_align_stat <- get_rnaseq_align_stat(input_dir)
```

## 1 Summary

```{r summary, results="asis"}
n_smp      <- length(rep_list)
total      <- df_align_stat %>% pull(total) %>% mean() %>% round(digits = 0)
map        <- df_align_stat %>% pull(map) %>% mean() %>% round(digits = 0)
map_pct    <- df_align_stat %>% pull(map_pct) %>% mean() %>% round(digits = 0)
unique_m   <- df_align_stat %>% pull(unique) %>% mean() %>% round(digits = 0)
unique_pct <- df_align_stat %>% pull(unique_pct) %>% mean() %>% round(digits = 0)
total <- round(total / 1e6, 1) # Million
map   <- round(map / 1e6, 1) # Million
unique_m <- round(unique_m / 1e6, 1) # Million
sum_text <- glue::glue("A total of {n_smp} samples; {total}M sequencing depth; ",
    "with {map}M ({map_pct}%) reads mapped to reference genome, including",
    "{unique_m}M ({unique_pct}%) unique mapped.")

print(sum_text)
```

## 2 Results

### 2.1 Table1. Total reads

```{r}
df_align_stat %>% to_DT
```

### 2.2 Figure 1. Number of mapped reads

+ Clean data percentage

```{r figure1_trim_reads, out.width = "100%", fig.height = 3}
rnaseq_trim_stat_plot(df_trim_stat)
```

+ Map percentage

```{r figure1_align_reads, out.width = "100%", fig.height = 3}
rnaseq_align_stat_plot(df_align_stat, 
                       columns = c("map", "unmap"),
                       title   = "Map reads")
```

+ Unique percentage

```{r figure1_unique_reads, out.width = "100%", fig.height = 3}
rnaseq_align_stat_plot(df_align_stat, 
                       columns = c("unique", "multiple", "unmap"),
                       title   = "Unique map reads")
```


### 2.3 RNAseq library type

Check the RNAseq library type, check if the reads from forward strand or reverse strand of mRNA;

Using `featureCounts` to count reads on mRNAs.

For strand-specific library

| sens | anti | library | example | 
| -- | -- | -- | -- |
| -s 1 | -s 2 | `read1: ++ -- / read2: -- ++` | - | 
| -s 2 | -s 1 | `read1: +- ++ / read2: +- -+` | dUTP library, NSR library |


```{r lib type, out.width = "100%"}
lib1 = "read1: ++ -- / read2: -- ++"
lib2 = "read1: +- ++ / read2: +- -+"
lib3 = "read1, read2, non-strand-specific"
strandness_sens <- get_rnaseq_strandness(input_dir)
if(strandness_sens == 0) {
  print(lib3)
} else if(strandness_sens == 1) {
  print(lib1)
} else if(strandness_sens == 2) {
  print(lib2)
} else {
  warning(paste0("unknown strandness, sens=", strandness_sens))
}
```


### 4. Quantification

```{r}
log2 <- glue::glue("Count reads on features:
                   Genome: {genome}
                      GTF: {basename(gtf)}")
print(log2)
```

**Sense strand**

```{r table2_sens_summary}
f_sum <- paste0(c(pd$args$count_sens, pd$args$count_anti), ".summary")
read_fc_summary(f_sum[1]) %>% dplyr::select(-sample) %>% dplyr::filter(count > 0) %>% to_DT
```

```{r count_sens, out.width = "100%", fig.height = 3}
read_fc_summary(f_sum[1]) %>% rnaseq_count_summary_plot
```

**Anti-sense strand**

```{r table2_anti_summary}
read_fc_summary(f_sum[2]) %>% dplyr::select(-sample) %>% dplyr::filter(count > 0) %>% to_DT
```

```{r, count_anti, out.width = "100%", fig.height = 3}
read_fc_summary(f_sum[2]) %>% rnaseq_count_summary_plot
```


**END**


