---
title: "Enrichment analysis"
author: "Ming Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
    keep_md: false
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  input_dir: ""
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(knitr))
knitr::opts_chunk$set(fig.width  = 12,
                      fig.height = 8,
                      fig.path   = "Figures/",
                      eval       = TRUE,
                      echo       = FALSE,
                      cache      = FALSE,
                      prompt     = FALSE,
                      tidy       = FALSE,
                      comment    = NA,
                      message    = FALSE,
                      warning    = FALSE,
                      eval       = T,
                      rownames.print = FALSE)
```


```{r load pkg}
library(hiseqr)
library(knitr)
library(ggplot2)
library(dplyr)
library(glue)
library(DT)
library(fishualize)
library(patchwork)
```


```{r config}
input_dir  <- params$input_dir
input_dir  <- "/data/yulab/wangming/work/devel_pipeline/hiseq/rnaseq/results/rnaseq_se/RNAseq_I_R_Hybrid_Dys_6h.vs.RNAseq_I_R_Hybrid_None_Dys_6h"
input_dir  <- normalizePath(input_dir)
if(! is_hiseq_dir(input_dir, "rnaseq_rx")) {
  stop(paste0("Not a RNAseq multiple directory, rnaseq_rx: ", input_dir))
}
```


```{r}
pd <- hiseqr::read_hiseq(input_dir)
enrich_dir <- list_hiseq_file(input_dir, "enrich_dir", "rnaseq_rx")
genome   <- list_hiseq_file(input_dir, "genome", "rnaseq_rx")
organism <- get_organism_name(genome)
arg_vars <- path_to_list(enrich_dir, recursive = TRUE)

# DEseq gene list
fix_xls <- get_fix_xls(input_dir)
df_sig  <- readr::read_delim(fix_xls, "\t") # sig, SYMBOL, required
if(! "SYMBOL" %in% names(df_sig)) {
  df_sig <- mutate(df_sig, SYMBOL = Gene)
}

df_gene <- df_sig %>%
  dplyr::mutate(link = gene_to_link(Gene, 
                                    organism = organism, 
                                    style    = "html", 
                                    readable = TRUE))
```


## 1 Summary

```{r}
msg <- glue::glue("Porject: {pd$args$project_name}; ",
                  "Control: {pd$args$wildtype_name}; ",
                  "Treatment: {pd$args$mutant_name}; ",
                  "Organism: {organism}; ",
                  "Genome_build: {pd$args$genome}")

print(msg)
```

Here are the numbers of significant changed genes, criteria: `foldChange >= 2, pvalue < 0.05`

### 1.1 Number of significant changed genes

```{r}
table(df_sig$sig) %>% as.data.frame %>% to_DT
```

```{r, fig.width = 5, fig.height = 5}
rnaseq_sig_stat_plot(df_sig)
```

### 1.2 Up-regulated genes

```{r}
df_gene %>%
  dplyr::filter(sig == "up") %>%
  dplyr::select(1:4, log2FoldChange, padj, sig) %>%
  dplyr::mutate(across(where(is.numeric), ~round(.x, digits = 2))) %>%
  to_DT(mode = 3)
```

### 1.3 Down-regulated genes

```{r}
df_gene %>%
  dplyr::filter(sig == "down") %>%
  dplyr::select(1:4, log2FoldChange, padj, sig) %>%
  dplyr::mutate(across(where(is.numeric), ~round(.x, digits =2 ))) %>%
  to_DT(mode = 3)
```



## 2 GO analysis

### 2.1 Up-regulated genes

```{r, fig.height = 6}
# wego plot
f1 <- pd$files$enrich$up$enrich_go$group_go.plot.wego.png
if(is.null(f1)) {
  print("GO() skipped")
} else if(file.exists(f1)) {
  knitr::include_graphics(f1)
}
```

```{r}
f1 <- pd$files$enrich$up$enrich_go$enrich_go.plot.BP.rds
f2 <- pd$files$enrich$up$enrich_go$enrich_go.plot.CC.rds
f3 <- pd$files$enrich$up$enrich_go$enrich_go.plot.MF.rds
plot_data <- lapply(c(f1, f2, f3), readRDS)
```

#### 2.1.1 GO enrichment: Dot plot

```{r, fig.width= 8, fig.height=12}
if(length(plot_data) > 0) {
  px <- lapply(1:3, function(i) {
    pi <- plot_data[[i]]
    pc <- pi$dotplot
    if(is(pc, "ggplot")) {
      pc
    }
  })
  px <- purrr::discard(px, is.null)
  patchwork::wrap_plots(px, ncol = 1)
}
```

#### 2.1.2 GO enrichment: enrich map plot

```{r, fig.width= 8, fig.height=15}
if(length(plot_data) > 0) {
  px <- lapply(1:3, function(i) {
    pi <- plot_data[[i]]
    pc <- pi$emapplot$plot_cluster
    if(is(pc, "ggplot")) {
      pc
    }
  })
  px <- purrr::discard(px, is.null)
  patchwork::wrap_plots(px, ncol = 1)
}
```

#### 2.1.3 GO enrichment: plot list

```{r}
# copy all png files 
up_dir <- file.path(enrich_dir, "up")
p_list <- list.files(up_dir, "*.png", all.files = TRUE, full.names = TRUE, 
                     recursive = TRUE)
p_list <- p_list[!grepl("kegg", p_list)]

p_df <- basename(p_list) %>%
  stringr::str_split("\\.", n = 2, simplify = TRUE) %>%
  as.data.frame %>%
  dplyr::rename(group = V1,
                plot  = V2) %>%
  dplyr::mutate(url = p_list) %>%
  dplyr::mutate(rel_link = file.path("data", group, basename(url)))
#--Function: copy files
file_copy <- function(x, y) {
  y_dir <- dirname(y)
  if(! dir.exists(y_dir)) {
    dir.create(y_dir, recursive = TRUE)
  }
  file.copy(x, y, copy.mode = TRUE, copy.date = TRUE)
}

# copy files
a = purrr::map2(p_df$url, p_df$rel_link, file_copy)
```


```{r}
#--Function: generate links
#--style: html, markdown
to_link <- function(name, url, style = "html") {
  if(style == "html") {
  paste0("<a href='", url, "' target='_blank'>", name, "</a>")
  } else if(style == "markdown") {
    paste0("[", name, "](", url, ")")
  } else {
    url
  }
}
# table
p_df2 <- p_df %>% 
  dplyr::mutate(link = to_link(basename(url), rel_link)) %>%
  dplyr::select(link) 

p_df2 %>% to_DT(mode = 3)
```

### 2.2 Down-regulated genes

```{r, fig.height = 6}
# wego plot
f1 <- pd$files$enrich$down$enrich_go$group_go.plot.wego.png
if(is.null(f1)) {
  print("GO() skipped")
} else if(file.exists(f1)) {
  knitr::include_graphics(f1)
}
```

```{r}
f1 <- pd$files$enrich$down$enrich_go$enrich_go.plot.BP.rds
f2 <- pd$files$enrich$down$enrich_go$enrich_go.plot.CC.rds
f3 <- pd$files$enrich$down$enrich_go$enrich_go.plot.MF.rds
plot_data <- lapply(c(f1, f2, f3), readRDS)
```

#### 2.2.1 GO enrichment: Dot plot

```{r, fig.width= 8, fig.height=12}
if(length(plot_data) > 0) {
  px <- lapply(1:3, function(i) {
    pi <- plot_data[[i]]
    pc <- pi$dotplot
    if(is(pc, "ggplot")) {
      pc
    }
  })
  px <- purrr::discard(px, is.null)
  patchwork::wrap_plots(px, ncol = 1)
}
```

#### 2.2.2 GO enrichment: enrich map plot

```{r, fig.width= 8, fig.height=15}
if(length(plot_data) > 0) {
  px <- lapply(1:3, function(i) {
    pi <- plot_data[[i]]
    pc <- pi$emapplot$plot_cluster
    if(is(pc, "ggplot")) {
      pc
    }
  })
  px <- purrr::discard(px, is.null)
  patchwork::wrap_plots(px, ncol = 1)
}
```

#### 2.2.3 GO enrichment: plot list

```{r}
# copy all png files 
down_dir <- file.path(enrich_dir, "down")
p_list <- list.files(down_dir, "*.png", all.files = TRUE, full.names = TRUE, 
                     recursive = TRUE)
p_list <- p_list[!grepl("kegg", p_list)]

p_df <- basename(p_list) %>%
  stringr::str_split("\\.", n = 2, simplify = TRUE) %>%
  as.data.frame %>%
  dplyr::rename(group = V1,
                plot  = V2) %>%
  dplyr::mutate(url = p_list) %>%
  dplyr::mutate(rel_link = file.path("data", group, basename(url)))
#--Function: copy files
file_copy <- function(x, y) {
  y_dir <- dirname(y)
  if(! dir.exists(y_dir)) {
    dir.create(y_dir, recursive = TRUE)
  }
  file.copy(x, y, copy.mode = TRUE, copy.date = TRUE)
}

# copy files
a = purrr::map2(p_df$url, p_df$rel_link, file_copy)
```


```{r}
#--Function: generate links
#--style: html, markdown
to_link <- function(name, url, style = "html") {
  if(style == "html") {
  paste0("<a href='", url, "' target='_blank'>", name, "</a>")
  } else if(style == "markdown") {
    paste0("[", name, "](", url, ")")
  } else {
    url
  }
}
# table
p_df2 <- p_df %>% 
  dplyr::mutate(link = to_link(basename(url), rel_link)) %>%
  dplyr::select(link) 

p_df2 %>% to_DT(mode = 3)
```


## 3 KEGG analysis

### 3.1 Up-regulated genes

```{r}
f <- pd$files$enrich$up$enrich_kegg$enrich_kegg.plot.rds
if(is.null(f)) {
  print("KEGG() skipped")
} else if(file.exists(f)) {
  plot_data <- readRDS(f)
}
```

#### 3.1.1 KEGG enrichment: Dot plot

```{r, fig.width= 8, fig.height=6}
plot_data$dotplot
```

#### 3.1.2 KEGG enrichment: enrich map plot

```{r, fig.width= 6, fig.height=6}
plot_data$emapplot$plot_cluster
```

#### 3.1.3 KEGG enrichment: plot list

```{r}
# copy all png files 
up_dir <- file.path(enrich_dir, "up")
p_list <- list.files(up_dir, "*.png", all.files = TRUE, full.names = TRUE, 
                     recursive = TRUE)
p_list <- p_list[grepl("kegg", p_list)]

p_df <- basename(p_list) %>%
  stringr::str_split("\\.", n = 2, simplify = TRUE) %>%
  as.data.frame %>%
  dplyr::rename(group = V1,
                plot  = V2) %>%
  dplyr::mutate(url = p_list) %>%
  dplyr::mutate(rel_link = file.path("data", group, basename(url)))
#--Function: copy files
file_copy <- function(x, y) {
  y_dir <- dirname(y)
  if(! dir.exists(y_dir)) {
    dir.create(y_dir, recursive = TRUE)
  }
  file.copy(x, y, copy.mode = TRUE, copy.date = TRUE)
}

# copy files
a <- purrr::map2(p_df$url, p_df$rel_link, file_copy)
```


```{r}
#--Function: generate links
#--style: html, markdown
to_link <- function(name, url, style = "html") {
  if(style == "html") {
  paste0("<a href='", url, "' target='_blank'>", name, "</a>")
  } else if(style == "markdown") {
    paste0("[", name, "](", url, ")")
  } else {
    url
  }
}
# table
p_df2 <- p_df %>% 
  dplyr::mutate(link = to_link(basename(url), rel_link)) %>%
  dplyr::select(link) 

p_df2 %>% to_DT(mode = 3)
```


### 3.2 Down-regulated genes

```{r}
f <- pd$files$enrich$down$enrich_kegg$enrich_kegg.plot.rds
if(is.null(f)) {
  print("KEGG() skipped")
} else if(file.exists(f)) {
  plot_data <- readRDS(f)
}
```

#### 3.2.1 KEGG enrichment: Dot plot

```{r, fig.width= 8, fig.height=6}
plot_data$dotplot
```

#### 3.2.2 KEGG enrichment: enrich map plot

```{r, fig.width= 8, fig.height=6}
plot_data$emapplot$plot_cluster
```

#### 3.2.3 KEGG enrichment: plot list

```{r}
# copy all png files 
down_dir <- file.path(enrich_dir, "down")
p_list <- list.files(down_dir, "*.png", all.files = TRUE, full.names = TRUE, 
                     recursive = TRUE)
p_list <- p_list[grepl("kegg", p_list)]

p_df <- basename(p_list) %>%
  stringr::str_split("\\.", n = 2, simplify = TRUE) %>%
  as.data.frame %>%
  dplyr::rename(group = V1,
                plot  = V2) %>%
  dplyr::mutate(url = p_list) %>%
  dplyr::mutate(rel_link = file.path("data", group, basename(url)))
#--Function: copy files
file_copy <- function(x, y) {
  y_dir <- dirname(y)
  if(! dir.exists(y_dir)) {
    dir.create(y_dir, recursive = TRUE)
  }
  file.copy(x, y, copy.mode = TRUE, copy.date = TRUE)
}

# copy files
a <- purrr::map2(p_df$url, p_df$rel_link, file_copy)
```


```{r}
#--Function: generate links
#--style: html, markdown
to_link <- function(name, url, style = "html") {
  if(style == "html") {
  paste0("<a href='", url, "' target='_blank'>", name, "</a>")
  } else if(style == "markdown") {
    paste0("[", name, "](", url, ")")
  } else {
    url
  }
}
# table
p_df2 <- p_df %>% 
  dplyr::mutate(link = to_link(basename(url), rel_link)) %>%
  dplyr::select(link) 

p_df2 %>% to_DT(mode = 3)
```


**END**
