---
title: "HiSeq Report"
author: "Ming Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
    keep_md: false
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  input_dir: ""
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(knitr))

knitr::opts_chunk$set(fig.width  = 12, 
                      fig.height = 8, 
                      fig.path   = "Figures/",
                      eval       = TRUE,
                      echo       = FALSE,
                      cache      = FALSE,
                      prompt     = FALSE,
                      tidy       = FALSE,
                      comment    = NA,
                      message    = FALSE,
                      warning    = FALSE,
                      eval       = T,
                      rownames.print = FALSE)
```


```{r}
library(hiseqr)
library(ggplot2)
library(patchwork)
library(fishualize)
library(dplyr)
library(glue)
library(ggcor)
library(GenomicRanges)
library(patchwork)
```


```{r config, eval = T}
input_dir  <- params$input_dir
input_dir  <- normalizePath(input_dir)
hiseq_type <- get_hiseq_type(input_dir)
if(! is_hiseq_multiple_dir(input_dir)) {
  stop(paste0("Not a RNAseq multiple directory, _rx: ", input_dir))
}
```


```{r}
pd       <- hiseqr::read_hiseq(input_dir)
organism <- pd$args$genome
genome   <- pd$args$genome
ctl_name <- pd$args$wildtype_name
exp_name <- pd$args$mutant_name
project_name <- pd$args$project_name

# smp merge list
merge_list <- list_rnaseq_merge_dirs(input_dir)
rep_list   <- list_rnaseq_single_dirs(input_dir)

# list of files
df_trim_stat  <- get_rnaseq_trim_stat(input_dir)
df_align_stat <- get_rnaseq_align_stat(input_dir)
report_list   <- get_rnaseq_report(input_dir)

# DEseq gene list
fix_xls <- get_fix_xls(input_dir)
df_sig  <- readr::read_delim(fix_xls, "\t") # sig, SYMBOL, required
if(! "SYMBOL" %in% names(df_sig)) {
  df_sig <- mutate(df_sig, SYMBOL = Gene)
}

df_gene <- df_sig %>%
  dplyr::mutate(link = gene_to_link(Gene, 
                        organism = organism, 
                        style    = "html", readable = TRUE))
```


## 1 Summary

```{r summary, results="asis"}
n_smp      <- length(rep_list)
total      <- df_align_stat %>% pull(total) %>% mean() %>% round(digits = 0)
map        <- df_align_stat %>% pull(map) %>% mean() %>% round(digits = 0)
map_pct    <- df_align_stat %>% pull(map_pct) %>% mean() %>% round(digits = 0)
unique_m   <- df_align_stat %>% pull(unique) %>% mean() %>% round(digits = 0)
unique_pct <- df_align_stat %>% pull(unique_pct) %>% mean() %>% round(digits = 0)
total <- round(total / 1e6, 1) # Million
map   <- round(map / 1e6, 1) # Million
unique_m <- round(unique_m / 1e6, 1) # Million
sum_text <- glue::glue("A total of {n_smp} samples; {total}M sequencing depth; ",
    "with {map}M ({map_pct}%) reads mapped to reference genome, including",
    "{unique_m}M ({unique_pct}%) unique mapped.")

print(sum_text)
```


## 2 Results

### 2.1 Table1. Total reads

```{r}
df_align_stat %>% to_DT
```

### 2.2 Figure 1. Number of mapped reads

+ Clean data percentage

```{r figure1_trim_reads, out.width = "100%", fig.height = 3}
rnaseq_trim_stat_plot(df_trim_stat)
```

+ Map percentage

```{r figure1_align_reads, out.width = "100%", fig.height = 3}
rnaseq_align_stat_plot(df_align_stat, 
                       columns = c("map", "unmap"),
                       title   = "Map reads")
```

+ Unique percentage

```{r figure1_unique_reads, out.width = "100%", fig.height = 3}
rnaseq_align_stat_plot(df_align_stat, 
                       columns = c("unique", "multiple", "unmap"),
                       title   = "Unique map reads")
```

## 2 Genes

### 2.1 Info

```{r}
msg1 <- glue::glue("Compare: log2( {exp_name} / {ctl_name} )
                    Project: {project_name}, 
                     Genome: {genome}
                    control: {ctl_name}
                  treatment: {exp_name}")
print(msg1)
```

### 2.2 Number of differentially expressed genes

Here are the numbers of significant changed genes, criteria: `foldChange >= 2, pvalue < 0.05`

```{r}
table(df_gene$sig) %>% as.data.frame %>% to_DT
```

```{r, fig.width = 4, fig.height = 4}
rnaseq_sig_stat_plot(df_gene)
```

```{r, fig.width=5, fig.height=4}
pub_rds <- pd$files$deseq$publish_plot_data.rds
if(is(pub_rds, "character")) {
  if(file.exists(pub_rds)) {
    dp <- readRDS(pub_rds)
    print(dp[1]) # scatter
    print(dp[3]) # volcano
  }
}
```

### 2.3 Up-regulated genes

```{r}
df_gene %>%
  dplyr::filter(sig == "up") %>%
  dplyr::select(1, link, 2:3, log2FoldChange, padj, sig) %>%
  dplyr::mutate(across(where(is.numeric), ~round(.x, digits =2 ))) %>%
  to_DT(mode = 3)
```


### 2.4 Down-regulated genes

```{r}
df_gene %>%
  dplyr::filter(sig == "down") %>%
  dplyr::select(1, link, 2:3, log2FoldChange, padj, sig) %>%
  dplyr::mutate(across(where(is.numeric), ~round(.x, digits =2 ))) %>%
  to_DT(mode = 3)
```



## 3 Detail report


```{r, report_list}
# copy files to current dir
data_dir <- "data"
dir.create(data_dir, showWarnings = FALSE)

if(length(report_list)) {
  local_html <- sapply(report_list, function(f) {
    f_prefix <- basename(dirname(dirname(f)))
    f_html   <- file.path(data_dir, paste0(f_prefix, ".rnaseq_report.html"))
    if(! file.exists(f_html)) {
      file.copy(f, f_html)
    }
    f_html
  })
} else {
  local_html = NULL
}
# show in table
if(length(local_html)) {
  data.frame(id = basename(dirname(dirname(report_list))),
             url = local_html
  ) %>%
    dplyr::mutate(link = paste0("<a href='", url, "' target='_blank'>", id, "</a>")) %>%
    dplyr::select(link) %>% 
    to_DT(mode = 3)
}
```


**END**


