---
title: "Enrichment analysis"
author: "Ming Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
    keep_md: false
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  deseq_dir: ""
  feature: "gene"
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(knitr))
knitr::opts_chunk$set(fig.width  = 12, 
                      fig.height = 8, 
                      fig.path   = "Figures/",
                      eval       = TRUE,
                      echo       = FALSE,
                      cache      = FALSE,
                      prompt     = FALSE,
                      tidy       = FALSE,
                      comment    = NA,
                      message    = FALSE,
                      warning    = FALSE,
                      eval       = T,
                      rownames.print = FALSE)
```


```{r load pkg}
library(hiseqr)
library(knitr)
library(ggplot2)
library(dplyr)
library(glue)
library(DT)
library(fishualize)
library(patchwork)
```


```{r config}
deseq_dir <- normalizePath(params$deseq_dir)
feature   <- params$feature # default: gene

# deseq_dir = "~/work/yu_2020/projects/20200528_lxh_RNAseq/result/RNAseq_read1/RNAseq_Harwich_w1118_6h.vs.RNAseq_w1118_Harwich_6h/"
# feature = "gene"
```


```{r}
# global vars
seq_type <- is_hiseq_dir(deseq_dir)
if(is.null(seq_type)) {
  stop("require deseq_single dir, eg: sampleA.vs.sampleB")
} else if(seq_type == "deseq_single") {
  args <- hiseqr::deseq_single_dir(deseq_dir)
} else {
  stop("require deseq_single directory, eg: sampleA.vs.sampleB")
}

pd <- hiseqr::read_deseq(deseq_dir, feature)#, ctl_vs_exp)
```


## 1 Summary

```{r, results="asis"}
# args <- pd$args
msg1 <- glue::glue("fold_change = log2( {args$prefix_exp} / {args$prefix_ctl} )")
print(msg1)
```


```{r}      
s <- glue::glue("Project:   {args$project_name}, 
                Reference: {args$genome}
                Feature:   {args$feature}
                control:   {args$prefix_ctl}
                treatment: {args$prefix_exp}")
print(s)
```

### 1.1 Number of genes

Here are the numbers of significant changed genes

```{r}
# DT::datatable(pd$count_df, rownames = FALSE, filter = "top", width = "40%",
#               options = list(dom = 't'))
knitr::kable(pd$count_df)
```

```{r, fig.width = 6, fig.height = 2}
pd$count_plot
```

### 1.2 Up-regulated genes

```{r}
sig_genes <- hiseqr::get_sig_gene(deseq_dir, feature) #, ctl_vs_exp)
df_up <- sig_genes$up %>%
  dplyr::mutate(symbol = hiseqr::gene_to_symbol(x = Gene, 
                                                organism = args$genome)) %>%
  dplyr::select(1:3, log2FoldChange, padj, sig, symbol) 

# DT::datatable(df_up, rownames = TRUE, filter = "top",
#               options = list(pageLength = 10, scrollX = TRUE))
get_DT_table(df_up)
```


### 1.3 Down-regulated genes

```{r}
df_down <- sig_genes$down %>%
  dplyr::mutate(symbol = hiseqr::gene_to_symbol(x = Gene, 
                                                organism = args$genome)) %>%
  dplyr::select(1:3, log2FoldChange, padj, sig, symbol) 

# DT::datatable(df_down, rownames = TRUE, filter = "top",
#               options = list(pageLength = 10, scrollX = TRUE))
get_DT_table(df_down)
```


## 2 GO analysis

### 2.1 up-regulated genes

```{r}
# data
f1    <- pd$file_list$enrich$up$go_plots.rds
pdata <- readRDS(f1)
```

#### 2.1.1 Overview of gene **Categories**

```{r, fig.width = 10, fig.height = 6}
pdata$group$p_wego
```

#### 2.1.2 Overview of **Enriched** genes

```{r, fig.width = 10, fig.height = 6}
pdata$enrich$p_wego
```

#### 2.1.3 Gene **Categories**

```{r, fig.width = 8, fig.height = 10, eval = FALSE}
# barplot
p1 <- pdata$group$BP$barplot
p2 <- pdata$group$CC$barplot
p3 <- pdata$group$MF$barplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.1.4 Enriched genes - dotplot

```{r, fig.width = 8, fig.height = 10}
# dotplot
p1 <- pdata$enrich$BP$dotplot
p2 <- pdata$enrich$CC$dotplot
p3 <- pdata$enrich$MF$dotplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.1.5 Enriched genes - cnetplot

```{r, fig.width = 8, fig.height = 10}
# dotplot
p1 <- pdata$enrich$BP$cnetplot
p2 <- pdata$enrich$CC$cnetplot
p3 <- pdata$enrich$MF$cnetplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.1.6 Enriched genes - emapplot

```{r, fig.width = 8, fig.height = 10, eval = FALSE, eval = FALSE}
# dotplot
p1 <- pdata$enrich$BP$emapplot  
p2 <- pdata$enrich$CC$emapplot
p3 <- pdata$enrich$MF$emapplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```



### 2.2 down-regulated genes

```{r}
# data
f2    <- pd$file_list$enrich$down$go_plots.rds
pdata <- readRDS(f2)
```

#### 2.2.1 Overview of gene **Categories**

```{r, fig.width = 10, fig.height = 6}
pdata$group$p_wego
```

#### 2.2.2 Overview of **Enriched** genes

```{r, fig.width = 10, fig.height = 6}
pdata$enrich$p_wego
```

#### 2.2.3 Gene **Categories**

```{r, fig.width = 8, fig.height = 10, eval = FALSE}
# barplot
p1 <- pdata$group$BP$barplot
p2 <- pdata$group$CC$barplot
p3 <- pdata$group$MF$barplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.2.4 Enriched genes - dotplot

```{r, fig.width = 8, fig.height = 10}
# dotplot
p1 <- pdata$enrich$BP$dotplot
p2 <- pdata$enrich$CC$dotplot
p3 <- pdata$enrich$MF$dotplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.2.5 Enriched genes - cnetplot

```{r, fig.width = 8, fig.height = 10}
# dotplot
p1 <- pdata$enrich$BP$cnetplot
p2 <- pdata$enrich$CC$cnetplot
p3 <- pdata$enrich$MF$cnetplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.2.6 Enriched genes - emapplot

```{r, fig.width = 8, fig.height = 10, eval = FALSE, eval = FALSE}
# dotplot
p1 <- pdata$enrich$BP$emapplot  
p2 <- pdata$enrich$CC$emapplot
p3 <- pdata$enrich$MF$emapplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```




### 2.3 Both up- and down- regulated genes

```{r}
# data
f3    <- pd$file_list$enrich$up_and_down$go_plots.rds
pdata <- readRDS(f3)
```

#### 2.3.1 Overview of gene **Categories**

```{r, fig.width = 10, fig.height = 6}
pdata$group$p_wego
```

#### 2.3.2 Overview of **Enriched** genes

```{r, fig.width = 10, fig.height = 6}
pdata$enrich$p_wego
```

#### 2.3.3 Gene **Categories**

```{r, fig.width = 8, fig.height = 10, eval = FALSE}
# barplot
p1 <- pdata$group$BP$barplot
p2 <- pdata$group$CC$barplot
p3 <- pdata$group$MF$barplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.3.4 Enriched genes - dotplot

```{r, fig.width = 8, fig.height = 10}
# dotplot
p1 <- pdata$enrich$BP$dotplot
p2 <- pdata$enrich$CC$dotplot
p3 <- pdata$enrich$MF$dotplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.3.5 Enriched genes - cnetplot

```{r, fig.width = 8, fig.height = 10}
# dotplot
p1 <- pdata$enrich$BP$cnetplot
p2 <- pdata$enrich$CC$cnetplot
p3 <- pdata$enrich$MF$cnetplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


#### 2.3.6 Enriched genes - emapplot

```{r, fig.width = 8, fig.height = 10, eval = FALSE, eval = FALSE}
# dotplot
p1 <- pdata$enrich$BP$emapplot  
p2 <- pdata$enrich$CC$emapplot
p3 <- pdata$enrich$MF$emapplot

# p1 / p2 / p3
cowplot::plot_grid(p1, p2, p3, ncol = 1)
```


## 3 KEGG analysis

### 3.1 up-regulated genes

```{r}
# data
f4    <- pd$file_list$enrich$up$kegg_plots.rds
pdata <- readRDS(f4)
```

#### 3.1.1 Overview of gene **Categories**

```{r, fig.width = 10, fig.height = 5}
pdata$enrich$p_wego
```


#### 3.1.2 Gene **Categories**

```{r, fig.width = 6, fig.height = 5, eval = FALSE}
# barplot
p1 <- pdata$enrich$kegg$barplot
p1
```


#### 3.1.3 Enriched genes - dotplot

```{r, fig.width = 6, fig.height = 5}
# dotplot
p1 <- pdata$enrich$kegg$dotplot
p1
```


#### 3.1.4 Enriched genes - cnetplot

```{r, fig.width = 6, fig.height = 6}
# cnetplot
p1 <- pdata$enrich$kegg$cnetplot
p1
```


#### 3.1.5 Enriched genes - emapplot

```{r, fig.width = 8, fig.height = 10, eval = FALSE, eval = FALSE}
# dotplot
p1 <- pdata$enrich$kegg$emapplot
p1
```



### 3.2 down-regulated genes

```{r}
# data
f5    <- pd$file_list$enrich$down$kegg_plots.rds
pdata <- readRDS(f5)
```

#### 3.2.1 Overview of gene **Categories**

```{r, fig.width = 10, fig.height = 5}
pdata$enrich$p_wego
```


#### 3.2.2 Gene **Categories**

```{r, fig.width = 6, fig.height = 5, eval = FALSE}
# barplot
p1 <- pdata$enrich$kegg$barplot
p1
```


#### 3.2.3 Enriched genes - dotplot

```{r, fig.width = 6, fig.height = 5}
# dotplot
p1 <- pdata$enrich$kegg$dotplot
p1
```


#### 3.2.4 Enriched genes - cnetplot

```{r, fig.width = 6, fig.height = 6}
# cnetplot
p1 <- pdata$enrich$kegg$cnetplot
p1
```


#### 3.2.5 Enriched genes - emapplot

```{r, fig.width = 8, fig.height = 10, eval = FALSE, eval = FALSE}
# dotplot
p1 <- pdata$enrich$kegg$emapplot
p1
```



### 3.3 up- and down- regulated genes

```{r}
# data
f6    <- pd$file_list$enrich$up_and_down$kegg_plots.rds
pdata <- readRDS(f6)
```

#### 3.3.1 Overview of gene **Categories**

```{r, fig.width = 10, fig.height = 5}
pdata$enrich$p_wego
```


#### 3.3.2 Gene **Categories**

```{r, fig.width = 6, fig.height = 5, eval = FALSE}
# barplot
p1 <- pdata$enrich$kegg$barplot
p1
```


#### 3.3.3 Enriched genes - dotplot

```{r, fig.width = 6, fig.height = 5}
# dotplot
p1 <- pdata$enrich$kegg$dotplot
p1
```


#### 3.3.4 Enriched genes - cnetplot

```{r, fig.width = 6, fig.height = 6}
# cnetplot
p1 <- pdata$enrich$kegg$cnetplot
p1
```


#### 3.3.5 Enriched genes - emapplot

```{r, fig.width = 8, fig.height = 10, eval = FALSE, eval = FALSE}
# dotplot
p1 <- pdata$enrich$kegg$emapplot
p1
```


**END**
