---
title: "HiSeq report"
author: "Ming Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
    keep_md: false
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  input_dir: ""
---


```{r setup, include=FALSE}
suppressPackageStartupMessages(library(knitr))

knitr::opts_chunk$set(fig.width  = 12, 
                      fig.height = 8, 
                      fig.path   = "Figures/",
                      echo       = FALSE,
                      cache      = FALSE,
                      prompt     = FALSE,
                      tidy       = FALSE,
                      comment    = NA,
                      message    = FALSE,
                      warning    = FALSE,
                      eval       = T,
                      rownames.print = FALSE)
```


```{r}
library(hiseqr)
library(ggplot2)
library(ggthemes)
library(ggrepel)
library(patchwork)
library(fishualize)
library(dplyr)
library(tidyr)
library(glue)
library(ggcor)
library(GenomicRanges)
```


```{r config, eval = T}
input_dir <- params$input_dir
input_dir <- normalizePath(input_dir)
if(! is_hiseq_dir(input_dir, "rx")) {
  stop(paste0("Input file is not a HiSeq rx directory: ", input_dir))
}
```


```{r}
pd <- summary_hiseq(input_dir, TRUE)
# summary
n_smp    <- length(list_hiseq_dir(input_dir, "r1"))
total    <- pd$trim %>% pull(input) %>% mean()
mito     <- pd$align %>% pull(chrM) %>% mean()
map      <- pd$align %>% pull(map) %>% mean()
mito_pct <- round(mito / total * 100, 2)
map_pct  <- round(map / total * 100, 2)
total <- round(total / 1e6, 2)
map   <- round(map / 1e6, 2)
mito  <- round(mito / 1e6, 2)
```


## 1 Summary

```{r summary, results="asis"}
sum_text <- glue::glue(
  "A total of {n_smp} samples; {total}M sequencing depth ",
  "with {map}M ({map_pct}%) reads mapped to reference genome; ",
  "and contains {mito}M ({mito_pct}%) Mitochondrial DNA reads."
)
print(sum_text)
```


## 2 Results

### 2.1 Table1. Total reads

```{r}
pd$align %>% 
  hiseqr::to_DT()
```

### 2.2 Figure 1. Number of mapped reads
+ Clean data percentage

```{r figure1_trim_reads, fig.width = 6}
plot_hiseq_trim(input_dir)
```

+ Map percentage

```{r figure1_align_reads, fig.width = 6}
plot_hiseq_align(input_dir, columns = c("map", "chrM", "spikein", "unmap"))
```

+ Unique percentage

```{r figure1_unique_reads, fig.width = 6}
plot_hiseq_align(input_dir, columns = c("unique", "multi", "unmap"))
```


### 2.3 Figure2. Number of peaks

```{r figure2_peaks, fig.width = 6}
p1 <- plot_hiseq_align(input_dir, 
                       columns = "unique", 
                       position = "identity", 
                       add_label = FALSE)
print(p1)
p2 <- plot_hiseq_peak(input_dir, hiseq_type = TRUE)
print(p2)
# patchwork::wrap_plots(p1, p2, ncol = 1)
```

### 2.4 FRiP.

> Fraction of reads in peaks (FRiP) – Fraction of all mapped reads that fall into the called peak regions, i.e. usable reads in significantly enriched peaks divided by all usable reads. In general, FRiP scores correlate positively with the number of regions. (Landt et al, Genome Research Sept. 2012, 22(9): 1813–1831)
>
> source: https://www.encodeproject.org/data-standards/terms/

> The fraction of reads in called peak regions (FRiP score) should be >0.3, though values greater than 0.2 are acceptable. For EN-TEx tissues, FRiP scores will not be enforced as QC metric. TSS enrichment remains in place as a key signal to noise measure
> source: https://www.encodeproject.org/atac-seq/


```{r table2_frip}
if(nrow(pd$frip) > 0) {
  pd$frip %>% 
    dplyr::select(name, total, map, n_peaks, pct) %>%
    hiseqr::to_DT()
}
```

### 2.5 Figure3. Fragment length

The insert size distribution of sequenced fragments from human chromatin had clear periodicity of approximately 200 bp, suggesting many fragments are protected by integer multiples of nucleosomes.

```{r figure3_fragment_length, out.width = "100%", fig.width = 6}
plot_hiseq_lendist(input_dir, hiseq_type = TRUE) + 
  theme(legend.position = "None")
```

### 2.6 Figure4. TSS enrichment

> Transcription Start Site (TSS) Enrichment Score - The TSS enrichment calculation is a signal to noise calculation. The reads around a reference set of TSSs are collected to form an aggregate distribution of reads centered on the TSSs and extending to 1000 bp in either direction (for a total of 2000bp). This distribution is then normalized by taking the average read depth in the 100 bps at each of the end flanks of the distribution (for a total of 200bp of averaged data) and calculating a fold change at each position over that average read depth. This means that the flanks should start at 1, and if there is high read signal at transcription start sites (highly open regions of the genome) there should be an increase in signal up to a peak in the middle. We take the signal value at the center of the distribution after this normalization as our TSS enrichment metric. Used to evaluate ATAC-seq. 
>
> source: https://www.encodeproject.org/data-standards/terms/

```{r figure4_tss_enrich, eval = T, fig.height = 4}
tss_list <- lapply(pd$enrich, function(i) {
  i$tss
}) %>% unlist()
tss_list <- purrr::keep(tss_list, file.exists)

gb_list <-  lapply(pd$enrich, function(i) {
  i$genebody
}) %>% unlist()
gb_list <- purrr::keep(gb_list, file.exists)

# TSS
if(length(tss_list) > 0) {
  knitr::include_graphics(tss_list)
}
if(length(gb_list) > 0) {
  knitr::include_graphics(gb_list)
}
```


### 2.7 Figure7. Correlation between replicates

The Pearson correlation coefficient were caculated, using the reads coverage on 500 bp window on the genome.

```{r fig7_bam_cor, eval = TRUE, fig.height = 4}
hm_list <- lapply(pd$cor, function(i) i$heatmap) %>% unlist()
hm_list <- purrr::keep(hm_list, file.exists)
if(length(hm_list) > 0) {
  knitr::include_graphics(hm_list)
}
```

#### PCA analysis for each replicate

```{r fig7_bam_pca, eval = TRUE, fig.height = 4}
pca_list <- lapply(pd$cor, function(i) i$pca) %>% unlist()
pca_list <- purrr::keep(pca_list, file.exists)
if(length(pca_list) > 0) {
  knitr::include_graphics(pca_list)
}
```

#### Overlap between peaks in each replicate

```{r fig7_peak_overlap, eval = TRUE, fig.height = 4}
ov_list <- lapply(pd$cor, function(i) {
  i$overlap
}) %>% unlist()
ov_list <- purrr::keep(ov_list, file.exists)
if(length(ov_list) > 0) {
  knitr::include_graphics(ov_list)
}
```



### 2.8 Figure 8. Bam fingerPrint

Answer the question: did the antibody-treatment enrich sufficiently so that the ChIP signal can be separated from the background signal? 

```{r fig8_bam_fingerprint, eval = TRUE, fig.height = 6}
bf_list <- lapply(pd$cor, function(i) i$fingerprint ) %>%
  unlist()
bf_list <- purrr::keep(bf_list, file.exists)
if(length(bf_list) > 0) {
  knitr::include_graphics(bf_list)
}
```



## 3 Detail report

```{r}
data_dir <- "data"
dir.create(data_dir, showWarnings = FALSE)

# from html
report_dirs <- list_hiseq_file(input_dir, "report_dir", TRUE)
report_html_list <- file.path(report_dirs, "HiSeq_report.html")

# to html
if(length(report_html_list)) {
  local_html <- sapply(report_html_list, function(f) {
    f_prefix <- basename(dirname(dirname(f)))
    f_html   <- file.path(data_dir, paste0(f_prefix, ".atac_report.html"))
    if(! file.exists(f_html)) {
      file.copy(f, f_html)
    }
    f_html
  })
} else {
  local_html <- NULL
}

# show in table
if(length(local_html)) {
  data.frame(id = basename(dirname(dirname(report_html_list))),
             url = local_html
  ) %>%
    dplyr::mutate(link = paste0("<a href='", url, "' target='_blank'>", id, "</a>")) %>%
    dplyr::select(link) %>%
    tibble::remove_rownames() %>% 
    to_DT(mode = 3)
}
```


## 4 ENCODE standard

> source: [https://www.encodeproject.org/atac-seq/#standards](https://www.encodeproject.org/atac-seq/#standards)

Experiments should have two or more biological replicates. Assays performed using EN-TEx samples may be exempted due to limited availability of experimental material, but at least two technical replicates are required.  

Each replicate should have 25 million non-duplicate, non-mitochondrial aligned reads for single-end sequencing and 50 million for paired-ended sequencing (i.e. 25 million fragments, regardless of sequencing run type). 

The alignment rate, or percentage of mapped reads, should be greater than 95%, though values >80% may be acceptable. 

Replicate concordance is measured by calculating IDR values (Irreproducible Discovery Rate). The experiment passes if both rescue and self consistency ratios are less than 2.

Library complexity is measured using the Non-Redundant Fraction (NRF) and PCR Bottlenecking Coefficients 1 and 2, or PBC1 and PBC2. The preferred values are as follows: NRF>0.9, PBC1>0.9, and PBC2>3. 

Various peak files must meet certain requirements. Please visit the section on output files under the pipeline overview for more information on peak files.

The number of peaks within a replicated peak file should be >150,000, though values >100,000 may be acceptable. 

The number of peaks within an IDR peak file should be >70,000, though values >50,000 may be acceptable.

A nucleosome free region (NFR) must be present.

A mononucleosome peak must be present in the fragment length distribution. These are reads that span a single nucleosome, so they are longer than 147 bp but shorter than 147*2 bp. Good ATAC-seq datasets have reads that span nucleosomes (which allows for calling nucleosome positions in addition to open regions of chromatin).

The fraction of reads in called peak regions (FRiP score) should be >0.3, though values greater than 0.2 are acceptable. For EN-TEx tissues, FRiP scores will not be enforced as QC metric. TSS enrichment remains in place as a key signal to noise measure.

Transcription start site (TSS) enrichment values are dependent on the reference files used; cutoff values for high quality data are listed in the table below. 

