---
title: "RNA-seq multiple report"
author: "Ming Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
    keep_md: true
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  input_dir: ""
  feature: "gene"
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(knitr))

knitr::opts_chunk$set(fig.width  = 12, 
                      fig.height = 8, 
                      fig.path   = "Figures/",
                      eval       = TRUE,
                      echo       = FALSE,
                      cache      = FALSE,
                      prompt     = FALSE,
                      tidy       = FALSE,
                      comment    = NA,
                      message    = FALSE,
                      warning    = FALSE,
                      eval       = T,
                      rownames.print = FALSE)
```


```{r load pkg}
library(hiseqr)
library(knitr)
library(ggplot2)
library(dplyr)
library(glue)
library(DT)
library(ggcor)
```


```{r config, eval = T}
# main 
feature    <- params$feature
input_dir  <- params$input_dir
input_dir  <- normalizePath(input_dir)
hiseq_type <- hiseqr::is_hiseq_dir(input_dir)

if(!hiseq_type == "rnaseq_multiple") {
  stop(paste0("Input dir is not RNAseq multiple directory: ", input_dir))
}
```


```{r global_vars}
args <- hiseqr::rnaseq_multiple_dir(input_dir, feature = feature)
```

## Summary

```{r summary, eval = T, results = "asis"}
# number of samples
n_samples <- length(args$smp_name)

# number of reads
n_total    <- sapply(args$single_args, function(f) ifelse(is.null(f[["reads_total"]]), 0, f[["reads_total"]]))
n_mito_u   <- sapply(args$single_args, function(f) ifelse(is.null(f[["mito.u"]]), 0, f[["mito.u"]]))
n_mito_m   <- sapply(args$single_args, function(f) ifelse(is.null(f[["mito.m"]]), 0, f[["mito.m"]]))
n_mito     <- n_mito_u + n_mito_m
n_genome_u <- sapply(args$single_args, function(f) ifelse(is.null(f[["genome.u"]]), 0, f[["genome.u"]]))
n_genome_m <- sapply(args$single_args, function(f) ifelse(is.null(f[["genome.m"]]), 0, f[["genome.m"]]))
n_unmap    <- sapply(args$single_args, function(f) ifelse(is.null(f[["unmap"]]), 0, f[["unmap"]]))
pct_genome_u <- round(sum(n_genome_u) / sum(n_total) * 100, 2)
pct_mito     <- round(sum(n_mito) / sum(n_total) * 100, 2)
pct_unmap    <- round(sum(n_unmap) / sum(n_total) * 100, 2)

# # number of reads
# n_total    <- mapply(function(f) f[["reads_total"]], args$single_args)
# n_mito_u   <- mapply(function(f) f[["mito.u"]], args$single_args)
# n_mito_m   <- mapply(function(f) f[["mito.m"]], args$single_args)
# n_mito     <- n_mito_u + n_mito_m
# n_genome_u <- mapply(function(f) f[["genome.u"]], args$single_args)
# n_genome_m <- mapply(function(f) f[["genome.m"]], args$single_args)
# n_unmap    <- mapply(function(f) f[["unmap"]], args$single_args)
# 
# pct_genome_u <- round(sum(n_genome_u) / sum(n_total) * 100, 2)
# pct_mito     <- round(sum(n_mito) / sum(n_total) * 100, 2)
# pct_unmap    <- round(sum(n_unmap) / sum(n_total) * 100, 2)

sum_text <- glue::glue("RNAseq multiple samples:
                       {n_samples} samples in all;
                       {sum(n_total)} (100%) reads in total, (min: {min(n_total)}, max: {max(n_total)});
                       {sum(n_genome_u)} ({pct_genome_u}%) reads map to genome, uniquely, (min: {min(n_genome_u)}, max: {max(n_genome_u)});
                       {sum(n_mito_u + n_mito_m)} ({pct_mito}%) reads map to rRNAs and tRNAs, (min: {min(n_mito_u + n_mito_m)}, max: {max(n_mito_u + n_mito_m)});
                       {sum(n_unmap)} ({pct_unmap}%) reads unmap, (min: {min(n_unmap)}, max: {max(n_unmap)});
                       ")
print(sum_text)
```

## Results


### 1 Table1. number of reads

```{r table1_mito_pct}
tmp <- lapply(args$single_args, function(f) f[["align_stat_df"]])
df  <- dplyr::bind_rows(tmp)

df_table <- df %>% 
  dplyr::group_by(id) %>%
  dplyr::mutate(total = sum(count)) %>%
  dplyr::mutate(group = forcats::fct_rev(group)) %>%
  tidyr::spread(group, count) %>%
  dplyr::mutate(genome_u_pct = round(genome.u / total * 100, 2))

# knitr::kable(df_table)
DT::datatable(df_table, rownames = FALSE, filter = "top", 
              options = list(pageLength = 10, scrollX = TRUE))
```


### 2 Figure 1. Number of mapped reads

```{r figure1_align_reads, out.width = "100%"}
alignPlot(df)
```


### 3. RNAseq library type

Check the RNAseq library type, check if the reads from forward strand or reverse strand of mRNA;

Using `featureCounts` to count reads on mRNAs.

arguments in `featureCounts`, `-s 1` indicates stranded, `-s 2` indicates reverse stranded.


```{r lib type, out.width = "100%"}
tmp1 <- sapply(args$single_args, function(f) f[["strandness_status"]])

# extract the pct number
align_stat_reader <- function(x){
  if(file.exists(x)){
    a  <- paste(readLines(x), collapse = ",")
    a1 <- stringr::str_match(a, "assigned_fwd:\\s+([0-9.]+%)")[, 2]
    a2 <- stringr::str_match(a, "assigned_rev:\\s+([0-9.]+%)")[, 2]
    ax <- c("fwd:-s=1" = a1, "rev:-s=2" = a2)
    return(ax)
  } else {
    return(c("fwd:-s=1"=0, "rev:-s=2"=0)) # blank
  }
}

df_lib_type <- data.frame(lapply(tmp1, align_stat_reader), stringsAsFactors = FALSE) %>%
  tibble::rownames_to_column("strand") %>%
  tidyr::gather("sample", "pct", -strand) %>%
  tidyr::spread(key = "strand", value = "pct") %>%
  tibble::rowid_to_column("number")


# knitr::kable(df_table)
DT::datatable(df_lib_type, rownames = FALSE, filter = "top", 
              options = list(pageLength = 10, scrollX = TRUE))
```


### 4. Quality control

+ PCA

+ Cor


**END**


