---
title: "Alignment report"
author: "Ming Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
    keep_md: true
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  input_dir: ""
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(knitr))

knitr::opts_chunk$set(fig.width  = 12, 
                      fig.height = 8, 
                      fig.path   = "Figures/",
                      eval       = TRUE,
                      echo       = FALSE,
                      cache      = FALSE,
                      prompt     = FALSE,
                      tidy       = FALSE,
                      comment    = NA,
                      message    = FALSE,
                      warning    = FALSE,
                      eval       = T,
                      rownames.print = FALSE)
```


```{r load pkg}
library(hiseqr)
library(knitr)
library(ggplot2)
library(dplyr)
library(glue)
library(DT)
# library(ggcor)
```


```{r config, eval = T}
# main 
input_dir  <- normalizePath(params$input_dir)

if(! hiseqr::is_align_dir(input_dir)) {
  stop(paste0("Input dir is not Align directory: ", input_dir))
}
```


```{r global_vars}
df_plot <- hiseqr::get_align_prj_stat(input_dir)

df_table <- df_plot %>%
  dplyr::mutate(group = forcats::fct_rev(group)) %>%
  tidyr::pivot_wider(names_from = group, values_from = count) %>% 
  dplyr::mutate(total = select(., -1) %>% rowSums()) %>%
  dplyr::mutate(map_pct = round(select(., starts_with("genome")) %>% rowSums() / total * 100, 1),
                mito_pct = round(select(., starts_with("mito")) %>% rowSums() / total * 100, 1))
```

## Summary

```{r summary, eval = T, results = "asis"}
n_smp <- length(unique(df_table$id))
total <- summary(df_table$total/1e6)
map   <- summary(df_table$map_pct)
mito  <- summary(df_table$mito_pct)
```

## Results


```{r }
msg <- glue::glue("A total of {n_smp} samples; average {round(total['Mean'], 1)}M reads for each sample were sequenced (min: {round(total['Min.'], 1)}M, max: {round(total['Max.'], 1)}M); \
    Map to genome {round(map['Mean'], 1)}%, and \
    rRNA {round(mito['Mean'], 1)}%.")

print(msg)

```


### 1 Table1. 

The `mito.m` and `mito.u` represent the reads mapped to rRNA.   

+ [NW_007931121.1](https://www.ncbi.nlm.nih.gov/nuccore/NW_007931121.1), Drosophila melanogaster chromosome X; Y rDNA sequence

+ [M21017.1](https://www.ncbi.nlm.nih.gov/nuccore/M21017.1/), D.melanogaster 18S, 5.8S 2S and 28S rRNA genes, complete, and 18S rRNA gene, 5' end, clone pDm238 

```{r table1_mito_pct}
# knitr::kable(df_table)
DT::datatable(df_table, rownames = FALSE, filter = "top", 
              options = list(pageLength = 10, scrollX = TRUE))
```

### 2 Figure 1. Number of mapped reads

```{r figure1_align_reads, out.width = "100%"}
hiseqr::align_plot(df_plot) %>%
  plotly::ggplotly()
```


**END**


